crates/vim/src/object.rs

```diff
@@ -812,6 +954,190 @@ mod test {
+        indoc! {"
+            ˇ
+            ˇ
+            ˇThe quick brown fox jumps
+            ˇover the lazy dog.
+
+
+            The quick brown fox jumps
+            over the lazy dog.
+        "},
+        // Newline paragraph and trailing newline paragraph
+        indoc! {"
+            ˇThe quick brown fox jumps
+            ˇover the lazy dog.
+            ˇ
+            ˇ
+            ˇ
+            ˇThe quick brown fox jumpsˇ
+            ˇover the lazy dog.ˇ
+            ˇ
+            ˇ
+            ˇ
+        "},
+        // "Empty" line paragraph with whitespace characters
+        indoc! {"
+            ˇThe quick brown fox jumps
+            ˇover the lazy dog.
+            ˇ
+            ˇ \t
+
+            ˇThe quick brown fox jumps
+            ˇover the lazy dog.ˇ
+            ˇ
+            ˇ \t
+            \t \t
+        "},
+        // Single line "paragraphs", where selection size might be zero.
+        indoc! {"
+            ˇThe quick brown fox jumps over the lazy dog.
+            ˇ
+            ˇThe quick brown fox jumpˇs over the lazy dog.ˇ
+            ˇ
+        "},
+    ];
+
+    #[gpui::test]
+    async fn test_change_paragraph_object(cx: &mut gpui::TestAppContext) {
+        let mut cx = NeovimBackedTestContext::new(cx).await;
+
+        for paragraph_example in PARAGRAPH_EXAMPLES {
+            cx.assert_binding_matches_all(["c", "i", "p"], paragraph_example)
+                .await;
+            cx.assert_binding_matches_all(["c", "a", "p"], paragraph_example)
+                .await;
+        }
+    }
+
+    #[gpui::test]
+    async fn test_delete_paragraph_object(cx: &mut gpui::TestAppContext) {
+        let mut cx = NeovimBackedTestContext::new(cx).await;
+
+        for paragraph_example in PARAGRAPH_EXAMPLES {
+            cx.assert_binding_matches_all(["d", "i", "p"], paragraph_example)
+                .await;
+            cx.assert_binding_matches_all(["d", "a", "p"], paragraph_example)
+                .await;
+        }
+    }
+
+    #[gpui::test]
+    async fn test_paragraph_object_with_trailing_normal_paragraph(cx: &mut gpui::TestAppContext) {
+        let mut cx = NeovimBackedTestContext::new(cx).await;
+
+        // Trailing normal paragraph
+        const TRAILING_NORMAL_PARAGRAPH_EXAMPLE: &'static str = indoc! {"
+
+
+            The quick brown fox jumps
+            over the lazy dog.
+            ˇ
+            ˇ
+            ˇThe quick brown fox jumpsˇ
+            ˇover the lazy dog.ˇ
+        "};
+
+        for keystrokes in [["c", "i", "p"], ["c", "a", "p"]] {
+            cx.assert_binding_matches_all(keystrokes, TRAILING_NORMAL_PARAGRAPH_EXAMPLE)
+                .await;
+        }
+        // Nvim in the test suite seems to be returning strange landing position.
+        cx.assert_binding_matches_all_exempted(
+            ["d", "i", "p"],
+            TRAILING_NORMAL_PARAGRAPH_EXAMPLE,
+            ExemptionFeatures::IncorrectLandingPosition,
+        )
+        .await;
```

Comment:
It looks like the tests are correctly emulating neovim here? It seems like maybe the difference arises depending on having a trailing newline on the buffer:

```
\n
ˇParagraph\n
```

In that case `d i p` will give you:

```
\n
ˇ\n
```

However if you have no trailing new line 

```
\n
ˇParagraph
```

vim gives you:

```
ˇ\n
```